lib/AST/LocalizationFormat.cpp

```diff
@@ -143,31 +132,46 @@ void YAMLLocalizationProducer::forEachAvailable(
   }
 }
 
+llvm::Optional<uint32_t> LocalizationInput::readID(llvm::yaml::IO &io) {
+  LocalDiagID diagID;
+  io.mapRequired("id", diagID);
+  if (diagID == LocalDiagID::NumDiags)
+    return llvm::None;
+  return static_cast<uint32_t>(diagID);
+}
+
 template <typename T, typename Context>
 typename std::enable_if<llvm::yaml::has_SequenceTraits<T>::value, void>::type
-readYAML(llvm::yaml::IO &io, T &Seq, bool, Context &Ctx) {
+readYAML(llvm::yaml::IO &io, T &Seq, T &unknownIDs, bool, Context &Ctx) {
   unsigned count = io.beginSequence();
-  if (count)
+  if (count) {
     Seq.resize(LocalDiagID::NumDiags);
+  }
+
   for (unsigned i = 0; i < count; ++i) {
     void *SaveInfo;
     if (io.preflightElement(i, SaveInfo)) {
-      DiagnosticNode current;
-      yamlize(io, current, true, Ctx);
-      io.postflightElement(SaveInfo);
+      io.beginMapping();
+      auto id = LocalizationInput::readID(io);
 
-      // A diagnostic ID might be present in YAML and not in `.def` file,
-      // if that's the case ScalarEnumerationTraits will assign the diagnostic ID
-      // to `LocalDiagID::NumDiags`. Since the diagnostic ID isn't available
-      // in `.def` it shouldn't be stored in the diagnostics array.
-      if (current.id != LocalDiagID::NumDiags) {
+      // If the current diagnostic ID is available in YAML and in `.def`, add it
+      // to the diagnostics array. Otherwise, re-parse the current diagnnostic
+      // id as a string and store it in `unknownIDs` array.
+      if (id) {
```

Comment:
Nit: could use s shortcut form of the if-exists - `if (auto id = LocalizationInput::readID(io)) {` since `id` is only useful inside of if condition.